Fix issue in enabling evented pleg feature gate

sairameshv opened the pull request:
#### What type of PR is this?
/kind bug
/kind flake

#### What this PR does / why we need it:
After enabling the evented pleg feature gate in the kubelet,
- GetNewerThan() [call](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/pod_workers.go#L1255) isn't blocking until the pod status/cache is updated and returning the empty pod status.
- Hence, whenever the `SyncLoop ADD/UPDATE/RECONCILE` functions are executed multiple times in a very less time interval, the Kubelet calls multiple `CreateContainer` CRI apis that results in the creation of duplicate containers within a given pod. 
- The initially created conainer keeps `Running` and the later container keeps `Exiting` and hence resulting the pod in `CrashLoopBackOff` state forever.

#### Which issue(s) this PR fixes:
Fixes https://github.com/kubernetes/kubernetes/issues/120941 https://github.com/kubernetes/kubernetes/issues/120140 https://github.com/kubernetes/kubernetes/issues/121003 

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fixed a regression in the Kubelet's behavior while creating a container when the `EventedPLEG` feature gate is enabled
```

sairameshv commented:
/cc @harche @mrunalp

sairameshv commented:
/test pull-kubernetes-node-crio-cgrpv1-evented-pleg-e2e

harche commented:
/test

harche commented:
/test ?

harche commented:
/test pull-kubernetes-node-crio-e2e

sairameshv commented:
/test pull-kubernetes-unit

bart0sh commented:
/triage accepted
/priority important-soon
/assign @vinaykul

harche commented:
/lgtm

harche commented:
/hold

harche commented:
/hold cancel

harche reviewed (commented):
/lgtm

harche commented:
flakiness of the [evented pleg node e2e job](https://testgrid.k8s.io/sig-node-presubmits#pr-crio-cgrpv1-evented-pleg-gce-e2e) is fixed with this PR.

harche commented:
> flakiness of the [evented pleg node e2e job](https://testgrid.k8s.io/sig-node-presubmits#pr-crio-cgrpv1-evented-pleg-gce-e2e) is fixed with this PR.

/cc @SergeyKanzhelev

mikebrow reviewed (commented):
see question

sairameshv commented on pkg/kubelet/container/cache.go:
```diff
@@ -102,15 +100,6 @@ func (c *cache) GetNewerThan(id types.UID, minTime time.Time) (*PodStatus, error
-		// Set the value in the cache only if it's not present already
-		// or the timestamp in the cache is older than the current update timestamp
-		if cachedVal, ok := c.pods[id]; ok && cachedVal.modified.After(timestamp) {
-			return false
```
I feel the cache needs to be set/updated (whenever the event has been received and processed) without any conditioning based on the timestamp. This keeps the pod cache data up to date.

mikebrow replied:
I figured this was to ensure old events were ignored that got queued up and processed after a newer get status request was processed..?  Note the comment block for the api "only if the data is newer than the cache"

sairameshv replied:
Addressed

sairameshv committed:
Fix issue in enabling evented pleg feature gate

Fixes https://github.com/kubernetes/kubernetes/issues/120941
GetNewerThan() call isn't blocking until the pod status/cache is updated and returning the empty pod status.
Hence, whenever the `SyncLoop ADD/UPDATE/RECONCILE` functions are called multiple times in a very less time interval,
Kubelet calls multiple `CreateContainer` CRI api that results in the creation of duplicate containers within a given pod.
The initially created conainer keeps `Running` and the later container keeps `Exiting` and hence resulting the pod in `CrashLoopBackOff` state forever

Signed-off-by: Sai Ramesh Vanka <svanka@redhat.com>

sairameshv commented:
Hey @mikebrow , Could you PTAL and give an LGTM tag on this PR?

saschagrunert reviewed (approved):
/lgtm

saschagrunert commented:
/test pull-kubernetes-e2e-kind

pacoxu commented:
Should we cherry-pick it to v1.26~v1.28? @sairameshv @saschagrunert

saschagrunert commented:
Yes, I'd be in favor of that. @sairameshv do you want to propose that or should I?

saschagrunert commented:
@sairameshv may I ask you to edit the release note section to be a bit more concrete?

SergeyKanzhelev commented:
KEP xref: https://github.com/kubernetes/enhancements/issues/3386

pacoxu commented:
> also fixes #121349

The failing test in https://testgrid.k8s.io/presubmits-kubernetes-nonblocking#pull-kubernetes-e2e-kind-alpha-features&width=90 is not fixed.

FYI

sairameshv commented:
> > also fixes #121349
> 
> The failing test in https://testgrid.k8s.io/presubmits-kubernetes-nonblocking#pull-kubernetes-e2e-kind-alpha-features&width=90 is not fixed.
> 
> FYI.
> 
> But currently, I still do not confirm if this is related to this issue. There are other failures there.

@pacoxu , I'm trying to actually figure out the issue from the e2e periodic job failures. But unfortunately, I couldn't find the related kubelet logs from the job artifacts(Ex: [evented-pleg job](https://testgrid.k8s.io/sig-node-cri-o#ci-crio-cgroupv1-evented-pleg))
Moreover, if I run the same job locally, not even once, it has failed.
So, I'm in the process of reproducing the issue.

pacoxu commented:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/121477/pull-kubernetes-e2e-kind-beta-features/1719488708551380992 is a general run that is based on #121477.
- The node cannot be ready for static pod issue. Faied in `Kubernetes e2e suite: [SynchronizedBeforeSuite]`
- There are two container log for static pods like scheduler/controller-manager, and one success and the other failed for port  in used.

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/121345/pull-kubernetes-e2e-kind-beta-features/1719259096076521472 is a run with disabled Evented PLGE FG.
- this run failed for other reasons: the node can be ready. The `SynchronizedBeforeSuite` passed.

I still
